### PR TITLE
Entity vtable fixes and new stuff

### PR DESCRIPTION
**Pull request by Mr-Auto**

### CODE REVIEW

File: src/game_api/sound_manager.hpp
Comment (Mr-Auto) [inline src/game_api/sound_manager.hpp:265]: Should i split this? (if yes, would love some name recommendations)

File: src/game_api/vtable_hook.hpp
Comment (Mr-Auto) [inline src/game_api/vtable_hook.hpp:58]: This is kind of bad, but also all the parameters in virtuals are just simple values and pointers, so there shouldn't be any big optimization issues.
Had to do this for reference parameters that won't work with std::move

File: src/game_api/rpc.cpp
Comment (Mr-Auto) [inline src/game_api/rpc.cpp:594]: Asked this on discord but didn't get replay
this changes the behavior slightly, but makes the most sense with the parameters given

File: src/game_api/entity.cpp
Comment (Mr-Auto) [inline src/game_api/entity.cpp:38]: wondering if this is just leftover from something else?

Comment (Dregu) [reply src/game_api/entity.cpp:38]: Maybe some old crap the UI used to need, maybe the remove_item wasn't there or maybe we wanted to really make sure we're definitely moving the topmost thing even though it should be detached already 🤷🏻‍♂️. These are like week 1 OL functions.

File: src/game_api/rpc.cpp
Comment (Dregu) [reply src/game_api/rpc.cpp:594]: So, did we think this virtual could be used to remove any entity from the items, but it actually just drops whatever is held when the uid is valid? Or did it have a different implementation that actually worked like it says, and was then simplified to a more broken one? Seems like an ok compromise anyway.

File: src/game_api/entities_items.hpp
Comment (Mr-Auto) [inline src/game_api/entities_items.hpp:760]: Better name suggestions? (since we already have `pick_up` function for movable

File: src/game_api/entity.cpp
Comment (Mr-Auto) [reply src/game_api/entity.cpp:38]: I will leave it as is but move it to `ui_util` since it's only used for UI

**Review by estebanfer — COMMENTED** [review]
Most of it is great! Just some small things to fix

File: src/game_api/entities_items.hpp
Comment (Mr-Auto) [reply src/game_api/entities_items.hpp:769]: no, this is only for all the `ENT_TYPE_ITEM_PICKUP_*`, it applies the effect of giving you ropes, power up etc.
I mostly want it as hookable function, for stuff like custom pickups

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (Mr-Auto) [reply src/game_api/script/usertypes/vtables_lua.cpp:121]: so maybe get rid of the "poof" part? or you have better name suggestion?

File: src/game_api/entities_monsters.hpp
Comment (Mr-Auto) [reply src/game_api/entities_monsters.hpp:381]: English is so stupid sometimes

File: src/game_api/movable.hpp
Comment (Mr-Auto) [reply src/game_api/movable.hpp:184]: Maybe, if anyone wants to test it (if it's used for anything else) and expose in future, go ahead

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (Mr-Auto) [reply src/game_api/script/usertypes/vtables_lua.cpp:120]: The comment is meaningless, it only means those entities have this virtual function implemented, but this function works for all entities, it is called on throw or drop etc.

File: src/game_api/entities_items.hpp
Comment (estebanfer) [reply src/game_api/entities_items.hpp:769]: what about uh, `get_powerup`, `give_powerup`, `grant_powerup`, `pick_powerup`, `consume_powerup`, `apply_powerup`?

File: src/game_api/movable.hpp
Comment (estebanfer) [reply src/game_api/movable.hpp:184]: So I guess nothing for this PR

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (estebanfer) [reply src/game_api/script/usertypes/vtables_lua.cpp:121]: `generate_fall_fx` or `generate_fall_effects`? Maybe add `floor_` before `fall`

File: src/game_api/entities_items.hpp
Comment (Mr-Auto) [reply src/game_api/entities_items.hpp:769]: I would not consider bomb bag a powerup, or like, not in current naming scheme that we or the game established
all of those suggestions sounds more like stand-alone functions rather then method on an object

Comment (Mr-Auto) [reply src/game_api/entities_items.hpp:769]: thou, if i had to i would probably choose `pick_powerup`

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (estebanfer) [reply src/game_api/script/usertypes/vtables_lua.cpp:120]: Oh so that comment is wrong, it says only for hired hands and lava pots

File: src/game_api/movable.hpp
Comment (Mr-Auto) [reply src/game_api/movable.hpp:184]: I just want to be done with this one, there is still work to do on exposing more, like the mount sounds that can be useful for making custom ones etc.

File: src/game_api/sound_manager.hpp
Comment (estebanfer) [reply src/game_api/sound_manager.hpp:267]: I guess it would be nice if it means more of those sound types exposed. But I don't really know about name, since now I guess we only have to use a name without altering the SoundMeta one...

Comment (Mr-Auto) [reply src/game_api/sound_manager.hpp:267]: You can edit the name, as long as there is no `new` constructor exposed to lua, no one should really care about the metatable name

File: src/game_api/entities_items.hpp
Comment (estebanfer) [reply src/game_api/entities_items.hpp:769]: > I would not consider bomb bag a powerup, or like, not in current naming scheme that we or the game established

I don't really see it like that from what I've seen, there's no POWERUP_BOMB or POWERUP_ROPE because there is just no need of such entities, and not because those aren't powerups (also I read in wikipedia, powerups like 1up or ammo are actually named powerup in there)

> all of those suggestions sounds more like stand-alone functions rather then method on an object

Fair, I would be fine with any other name as long as it's distinguishable enough from the `pick_up` functions that's just grabbing something your hands. Though for me it seems fine to me to see some code like `pickup:give_powerup(player)`, because it means giving the pickup's "contained" powerup to the player

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (estebanfer) [reply src/game_api/script/usertypes/vtables_lua.cpp:120]: I still would prefer some comment so the people that see it in the API docs know better what it means without having to try it themselves, but I leave it up to you, resolve the convo or add some comment

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (estebanfer) [reply src/game_api/script/usertypes/vtables_lua.cpp:121]: `generate_landing_effects`, maybe `on_landing` if it's not only effects perhaps?

File: src/game_api/entities_items.hpp
Comment (Mr-Auto) [inline src/game_api/entities_items.hpp:105]: Btw. no objections to this name? I didn't like the "buy" name since it does not deal with the money and the ownership is janky as for the comment

File: src/game_api/entities_items.hpp
Comment (estebanfer) [reply src/game_api/entities_items.hpp]: I guess I'll mark this as resolved now

File: src/game_api/entities_items.hpp
Comment (estebanfer) [reply src/game_api/entities_items.hpp:105]: ah, I see, seems to be the function that gets called after the item is bought, to make the player to grab the item if possible (grab item, equip backitem), the comment for me made it seem like its a side effect of the function, rather than it's main thing.
And for the name I'm not sure, I guess it's fine as long as the doc comments is good enough? I think it's a little ambiguous but I'm not gonna be perfectionist, it could be fine as that if you also think so, and we don't come up with any better name

Comment (estebanfer) [reply src/game_api/entities_items.hpp:105]: Testing a bit more, for powerup pickupables, calling it directly seems to just grant the powerup to the entity, just like being bought for free, same with backitems. But for items like guns, it just makes the player to grab the entity without doing anything else

File: docs/src/includes/_types.md
Comment (estebanfer) [inline docs/src/includes/_types.md:4248]: I forgot to say, the `set_draw_depth` function now requires an unknown b3f param, I think it would be fine, if anyone could have any idea what's b3f when using it. I know the old function is still usable, but it's NoDoc.
I think the best option is to either say what is b3f if you know, make b3f optional or else make the other function visible

File: src/game_api/entities_items.hpp
Comment (Mr-Auto) [reply src/game_api/entities_items.hpp:105]: If i remember from testing, it removed the gun from the shop items.

File: docs/src/includes/_types.md
Comment (Mr-Auto) [reply docs/src/includes/_types.md:4248]: Well, the param is optional, just not shown as optional in the doc, i can try change that

I have no idea that's the purpose of it, game does something with it when going from one layer to another. But also there is default value in EntityDB, so no clue at all

Comment (estebanfer) [reply docs/src/includes/_types.md:4248]: > just not shown as optional in the doc, i can try change that
Yeah, that was my issue, the `spel2.lua` declarations will now complain every time you skip the parameter (as it already does with other things since some time, I should fix them also later). Also new people using the API would think it's required it the documentation says so

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (Mr-Auto) [reply src/game_api/script/usertypes/vtables_lua.cpp:121]: `generate_floor_landing_effects` ? fell like the floor is important here, since it does not work on stomping enemies, which can also be considered "landing on something". Or maybe just explain it in the comment since it's kind of long?

I still need to add comment that it does not generate the damage sounds or floor sounds when you fall while being stunned etc. (thrown by something for example)

File: src/game_api/entities_items.hpp
Comment (Mr-Auto) [reply src/game_api/entities_items.hpp:105]: ok, so ... I think it's still valid function to detect purchase. Thou needs better comment and maybe better name?

File: docs/src/includes/_types.md
Comment (Mr-Auto) [reply docs/src/includes/_types.md]: Changed it, the comment should show up after #399 is merged

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (estebanfer) [reply src/game_api/script/usertypes/vtables_lua.cpp:121]: Either the long name or comment would work. Landing alone may not be clear enough if stomping counts, but it's probably enough to make someone doubt if it does and check the comment

File: src/game_api/entities_items.hpp
Comment (estebanfer) [reply src/game_api/entities_items.hpp:105]: if you don't wanna include the purchase action in it: `equip`, `try_equip`, `try_auto_equip`

Comment (Mr-Auto) [reply src/game_api/entities_items.hpp:105]: the `on_` are usually problematic, as the hooks for me look stupid as `set_pre_on_something`

File: src/game_api/vtable_hook.hpp
Comment (Mr-Auto) [reply src/game_api/vtable_hook.hpp]: Reverted this for now, it's kind of nice to have parameters that use reference and not pointer, to show that game does not do null check for it, but i guess it doesn't matter much since it is not shown in the API doc

File: src/game_api/sound_manager.hpp
Comment (estebanfer) [reply src/game_api/sound_manager.hpp:267]: Honestly, I can't come up with any name ideas, since I'm not completely sure what's the difference between those clases :cat2:.
Is the current SoundMeta for loop-able long sounds? Having a bool for `play_ending_secuence` makes it sound like it's repeatable and can be ended by game logic, and also the `start_over`

Comment (estebanfer) [reply src/game_api/sound_manager.hpp:267]: I guess you could leave it for another PR if you wanted

Comment (Mr-Auto) [reply src/game_api/sound_manager.hpp:267]: The `start_over` is more of a name i give it to make sense in the API rather then what actual functionality it holds.
The class name doesn't need to reflect much. In the source code it's probably just base class like "Sound".

That base class is used for stuff like: play and forget. You can just play sfx and not worry about it as the internals of it will deal with cleaning after it finishes etc.
Then `SoundMeda` is for sounds you want you either reuse or need to maintain (like keep updating the position etc.), it also used when you need to know if previous sound is still playing. Pretty sure you need to call the destroy function on it yourself

At the end is `BackgroundSound` which is probably only implemented for the main "music" in game_manager

There might be more that we don't know of, or that don't differ by the class members (only virtual overrides)

Comment (estebanfer) [reply src/game_api/sound_manager.hpp:267]: some rough ideas, `ContinuousSound`, `LoopableSound`? I suppose some examples could be the UFO laser shoot, lamassu laser prepare sound or Lamahu tentacles.
Do you have any ideas for the name yourself?

File: src/game_api/script/usertypes/vtables_lua.cpp
Comment (Mr-Auto) [reply src/game_api/script/usertypes/vtables_lua.cpp:197]: yes xd

File: src/game_api/sound_manager.hpp
Comment (Mr-Auto) [reply src/game_api/sound_manager.hpp:267]: SFX -> Sound -> BackgroundSound  ?

Comment (estebanfer) [reply src/game_api/sound_manager.hpp:267]: Sounds pretty good
It could also be named SoundEffect and Sound changed into something like DynamicSound/ActiveSound to give contrast with the base class I guess
Also, is this gonna continue on this PR or should I approve this after testing everything works?

Comment (Mr-Auto) [reply src/game_api/sound_manager.hpp:267]: > Also, is this gonna continue on this PR or should I approve this after testing everything works?

We can move this conversation to discord and be done with this PR i guess

File: src/game_api/entities_items.hpp
Comment (estebanfer) [inline src/game_api/entities_items.hpp:769]: Well uh, now I found an issue here, skipping it doesn't seem to make pickups unpickable, and the boolean parameter seems to be for playing the sound or not (calling it with true plays the sound, with false it doesn't play). Probably named something like `play_sound` or `play_sfx`

Comment (estebanfer) [reply src/game_api/entities_items.hpp:769]: Skipping it does prevent anything from happening when using `E:equip(P)` though

File: src/game_api/sound_manager.hpp
Comment (estebanfer) [reply src/game_api/sound_manager.hpp:267]: Alright

File: src/game_api/entities_items.hpp
Comment (Mr-Auto) [reply src/game_api/entities_items.hpp:769]: fixed

Comment (estebanfer) [reply src/game_api/entities_items.hpp:769]: no name given for the bool parameter?

Comment (Mr-Auto) [reply src/game_api/entities_items.hpp]: i forgor

**Review by estebanfer — APPROVED** [review]
Approved! (finally)

**Pull request merged**
